rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,370 @@
+CREATE INDEX hour_of_day ON hudi_table (DATE_FORMAT(timestamp, '%H'));
+```
+
+2. Write config in ingestion: A new write config is introduced `hoodie.write.index.function.list` to allow a user to add
+   and update Index Functions.
+
+Example:
+
+```
+hoodie.write.index.function.list=timestamp,datestr,col_stats,DATE_FORMAT(timestamp, '%Y-%m-%d');timestamp,,,DATE_FORMAT(timestamp, '%H')
+```
+
+#### Index in Metadata Table
+
+For each Index Function, Hudi needs to populate the index metadata to facilitate file pruning. We use the metadata table
+to store these index metadata. The index metadata of each individual Index Function is stored under a new partition with
+the naming of the Index Function name and the index type, e.g., `datestr_col_stats`. For the default index type of
+file-level column stats, the index format and logic of existing column stats index is reused. Once an Index Function is
+created and the corresponding index metadata is populated in metadata table, the index metadata is always consistent
+with the data table during write transactions, using the same multi-table transaction for updating metadata table.
+
+We plan to also integrate Index Function with Hudi's Asynchronous Indexing so that the index metadata for an Index
+Function can be populated in an async way.
+
+#### File Pruning
+
+When a query with the predicates related to an Index Functions is being planned, Hudi intercepts such predicate to carry
+out the file pruning
+
+1. Identify the relevant Index Function based on the predicate and add a new predicate based on the Index Function name
+   if needed. Take the Index Function created by
+   SQL `CREATE INDEX datestr ON hudi_table USING col_stats (DATE_FORMAT(timestamp, '%Y-%m-%d'));`:
+
+- If the original data column of an Index Function is used, transform the predicate to a new predicate based on the
+  Index Function name if possible, e.g., `timestamp between 1666767600 and 1666897200`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the expression of an Index Function is used, replace the expression to the Index Function name,
+  e.g., `DATE_FORMAT(timestamp, '%Y-%m-%d') between '2022-10-26' and '2022-10-27'`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the Index Function name is used in the predicate, e.g., `datestr between '2022-10-26' and '2022-10-27'`, no action
+  is needed here.
+
+2. For the Index Function name appeared in the predicate, look up the corresponding index metadata from metadata table
+   and prune files that are not needed for scanning.
+
+For each query engine, we need to intercept the engine-specific representation of the predicates, and derive the
+relevant predicate based on the function. We consider integrating Spark and Presto first in the implementation.
+
+#### Index-Function-Based Ordering on Write Path
+
+To unlock efficient file pruning, we plan to support Index-Function-based ordering during write operations. Similar to
```

Comment:
I'd suggest we reframe it a bit differently: essentially we don't want to introduce a new sorting mechanism (there's already bulk-insert sorting mode, clustering), instead we should clarify that these derived columns should be usable by both bulk-insert sorting mode and clustering so that people can control whether they want to adjust table's layout based on these.